### PR TITLE
Initial support for custom mod world types

### PR DESCRIPTION
**Pull request by Niko-sk2x**
~~This is mostly work in progress. It works but there may be some debug code that slipped through, some minor things may not work perfectly, and some parts are going to be changed even when they are fine right now.

Reviews are welcome, even in this work in progress state.~~

It should be done now.

### CODE REVIEW

File: src/main/java/cubicchunks/CommonEventHandler.java [hunk 1]
```diff
@@ -62,16 +60,26 @@
 @ParametersAreNonnullByDefault
 @MethodsReturnNonnullByDefault
 public class CommonEventHandler {
+    public static final String IS_CUBIC_CHUNKS = "icCubicChunks";
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/CommonEventHandler.java]: Typo.

File: src/main/java/cubicchunks/CommonEventHandler.java [hunk 2]
```diff
@@ -84,16 +92,15 @@ public void onWorldAttachCapabilities(AttachCapabilitiesEvent<World> evt) {
         if (type instanceof ICubicWorldType) {
             generationRange = ((ICubicWorldType) type).calculateGenerationHeightRange((WorldServer) world);
         }
-        WorldSavedDataHeightBounds heightBounds =
-                (WorldSavedDataHeightBounds) evt.getObject().getMapStorage().getOrLoadData(WorldSavedDataHeightBounds.class, "heightBounds");
-        if (heightBounds == null) {
-            heightBounds = new WorldSavedDataHeightBounds("heightBounds");
+
+        if (savedData == null) {
+            savedData = new WorldSavedCubicChunksData("heightBounds");
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/CommonEventHandler.java]: Probably should be ```new WorldSavedCubicChunksData("cubicChunksData");```

File: src/main/java/cubicchunks/CommonEventHandler.java [hunk 3]
```diff
@@ -84,16 +92,15 @@ public void onWorldAttachCapabilities(AttachCapabilitiesEvent<World> evt) {
...
         }
-        int minHeight = heightBounds.minHeight;
-        int maxHeight = heightBounds.maxHeight;
+        int minHeight = savedData.minHeight;
+        int maxHeight = savedData.maxHeight;
         world.initCubicWorldServer(new IntRange(minHeight, maxHeight), generationRange);
-        heightBounds.markDirty();
-        evt.getObject().getMapStorage().setData("heightBounds", heightBounds);
+        savedData.markDirty();
+        evt.getObject().getMapStorage().setData("heightBounds", savedData);
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/CommonEventHandler.java]: Same here?

File: src/main/java/cubicchunks/CubicChunks.java [hunk 4]
```diff
@@ -330,7 +330,10 @@ public int getValue() {
             USE_VANILLA_CHUNK_WORLD_GENERATORS(false,
                     "Enabling this option will force " + CubicChunks.MODID
                             + " to use world generators designed for two dimensional chunks, which are often used for custom ore generators added by mods. To do so "
-                            + CubicChunks.MODID + " will pregenerate cubes in a range of height from 0 to 255.");
+                            + CubicChunks.MODID + " will pregenerate cubes in a range of height from 0 to 255."),
+            FORCE_CUBIC_CHUNKS(false,
+                    "Enabling this will force creating a cubic chunks world, even if it's not cubic chunks world type. This option is automatically"
+                            + " set in world creation GUI when creating cubic chunks world with non-cubicchunks world type");
```
Comment (Foghrye4) [inline src/main/java/cubicchunks/CubicChunks.java:336]: I don't get it.

> This option is automatically set in world creation GUI when creating cubic chunks world with non-cubicchunks world type

Do you mean: "If set to *true*, launching custom non-cubicchunks world creation settings GUI will show options available for cubic chunks worlds"?

File: src/main/java/cubicchunks/CommonEventHandler.java (hunk 2 above)
Comment (Niko-sk2x) [reply src/main/java/cubicchunks/CommonEventHandler.java]: I wanted to change that and forgot.

File: src/main/java/cubicchunks/CubicChunks.java (hunk 4 above)
Comment (Niko-sk2x) [reply src/main/java/cubicchunks/CubicChunks.java:336]: No, it's literally an option changed by the button in world creation GUI. If you set it to true, you will have new cubic chunks worlds by default, unless you opt out. If you set it to false, you will have vanilla worlds by default unless you opt in.

This also allows to create cubic chunks worlds serversde if you set it to true, as there is no GUI there.

**Pull request merged**